draft-bormann-core-responses.md

```diff
@@ -178,6 +257,35 @@ This contains an opaque string with the port number as a 16-bit number
 |-----|---|---|---|---|------------|--------|--------|---------|
 | TBD | C | U | - | - | Respond-To | opaque |   6-18 | (none)  |
 
+
+## Leisure-For-Responses
+
+This new option (elective, proxy-unsafe, uint valued) allows the server
+to send up as many responses as indicated in the option value in
+addition to the requested response. They are to be sent immediately
+after the original response.
+
+On its own, the option does not indicate which more responses the client
+would expect (though further elective proxy-safe no-cache-key options
+can be added on top of that to give better guidance), and the server may
+choose not to send any at all.
+
+Intermediaries may add or remove the option, and use incoming responses to
```

Comment:
Yes, as are all options that allow non-traditional responses (Observe and Block2 explicitly, requests to multicast addresses in their original versions ~"need some extra stuff out of scope of these documents", and the upcoming Multicast-Signaling explicitly again). Any such option requires the client to use a token that outlives the original response/'s exchange lifetime, and any client end of a proxy needs to be aware of that.

That is a reason why we may prefer to have a Leisure-For and an extra option that allows sending, say, multiple blocks or follow-up data: In such a split, only Leisure-For would need to be hop-by-hop, and the other could be safe-to-forward (but a proxy that understands it may thin out responses like it may with Observe). But there are sure also good reasons against silently introducing CONNECT-lite ;-)

Reply:
For a concrete example on why it has to be, assume a client sends it to allow the server to send multiple response blocks.

The proxy would receive several blocks; in the likely case it receives the initial (the requested) block first, stores it in its cache and forwards it; later responses don't match a pending request and are discarded. But if a later block arrives first, what'll the proxy do? If it looks at the descriptive-use Block2 block number, it considers the response wrong, won't populate the cache and send the client a generic "proxy error". Worse, if it doesn't look at it (7959 says it MUST take not of the block size and MUST check ETags, but makes no requirement to verify the block number, at least on a quick check), it puts the data into the wrong cache part and forwards that.

So yeah, no improvements over the existing cases, still hop-by-hop.